device_doctor/lib/src/ios_debug_symbol_doctor.dart

```diff
@@ -0,0 +1,188 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io' as io;
+
+import 'package:args/command_runner.dart';
+import 'package:file/file.dart';
+import 'package:file/local.dart';
+import 'package:process/process.dart';
+import 'package:logging/logging.dart';
+
+class DiagnoseCommand extends Command<bool> {
+  DiagnoseCommand({
+    this.processManager = const LocalProcessManager(),
+    Logger? loggerOverride,
+  }) : logger = loggerOverride ?? Logger.root;
+
+  final Logger logger;
+
+  final ProcessManager processManager;
+
+  final String name = 'diagnose';
+  final String description = 'Diagnose whether attached iOS devices have errors.';
+
+  Future<bool> run() async {
+    final List<String> command = <String>['xcrun', 'xcdevice', 'list'];
+    final io.ProcessResult result = await processManager.run(
+      command,
+    );
+    if (result.exitCode != 0) {
+      logger.severe(
+        '$command failed with exit code ${result.exitCode}\n${result.stderr}',
+      );
+      return false;
+    }
+    final Iterable<XCDevice> devices = XCDevice.parseJson(result.stdout as String);
+    final Iterable<XCDevice> devicesWithErrors = devices.where((XCDevice device) => device.hasError);
+
+    if (devicesWithErrors.isNotEmpty) {
+      logger.severe('Found devices with errors!');
+
+      for (final XCDevice device in devicesWithErrors) {
+        logger.severe('${device.name}: ${device.error}');
+      }
+      return false;
+    }
+
+    return true;
+  }
+}
+
+class RecoverCommand extends Command<bool> {
+  RecoverCommand({
+    this.processManager = const LocalProcessManager(),
+    Logger? loggerOverride,
+    this.fs = const LocalFileSystem(),
+  }) : logger = loggerOverride ?? Logger.root {
+    argParser
+      ..addOption(
+        'cocoon-root',
+        help: 'Path to the root of the Cocoon repo. This is used to find the Build dashboard macos project, which is '
+            'then used to open Xcode.',
+        mandatory: true,
+      )
+      ..addOption(
+        'timeout',
+        help: 'Integer number of seconds to allow Xcode to run before killing it.',
+        defaultsTo: '300',
```

Comment:
I mean locally, it syncs within 30 seconds. I thought 5 minutes was a conservatively long interval. I can update the default if you like; however, I was planning for this recover command to be triggered via led, and having the recipe read from properties for the timeout value. WDYT?

Reply:
A platform level property SGTM. We can use smaller numbers there to avoid increasing exec time too much (hopefully the XCdevice error doesn't happen too frequently).